Examples: Add TSL Procedural Wood Material

SeeleyLogan opened the pull request:
**Description**

This pull requests extends the ThreeJS examples library and adds an addon to JSM.

The addon provides a material generator for procedural wood textures. The addon includes preset wood parameters that can easily be exposed and changed.

The example creates a scene demonstrating each wood preset imported from the addon.

The math behind the wood texture is ported from a [blender tutorial](https://www.youtube.com/watch?v=n7e0vxgBS8A) by [Lance Phan](https://www.youtube.com/@LancePhan).

Creating procedural materials can be time consuming. This contribution will allow developers to skip lower-level wood material design; instead focusing on high level customization.

![webgpu_procedural_wood](https://github.com/user-attachments/assets/5f0689b0-126d-487f-a06d-3a90f0cffa15)

**Caveat**

Each wood texture has the same node graph; different parameters. This caused ThreeJS to cache the first shader and ignore the parameters. I avoided this issue by bypassing the caching altogether, causing my example to compile each material individually (regardless of its parameters). This shouldn't be necessary and I believe I am doing something wrong on that end.

*This contribution is funded by [DriveCore](https://drivecore.ai)*

SeeleyLogan committed:
Added webgpu_procedural_wood

(cherry picked from commit fb0d8832c2e703e01998e235b8604504e909dd3c)

SeeleyLogan committed:
Robustness

(cherry picked from commit 5711d8f977e8d737a0c5c17f3c6dc74b4548c979)

SeeleyLogan committed:
Fixed Cedar

(cherry picked from commit d1813a26467f64a5bb4340e20f2d24823648b41e)

bhouston committed:
rename example, change description

(cherry picked from commit 2a3572593c06cd634feaf0dfbc3030b377371373)

sunag commented:
>  @sunag do you have an idea on how to make this performant given they are all basically the same material, just different parameters?

My first suggestion is to use uniforms instead of constants for variable parameters, I didn't find any uniforms in the code.

The code also needs to follow the mrdoob code style, I would also suggest creating an extended material from `MeshPhysicalNodeMaterial` instead of the `GenerateWoodMaterial` function.

I'm not sure if that would be enough, but I can continue the analysis as we progress.

SeeleyLogan committed:
Update ProceduralWood.js

(cherry picked from commit 9ae3876a8313206c82668abf236d507ee186f6ca)

sunag commented:
> @sunag I'm having trouble with these uniforms. The value of the uniforms need to be different per-material... If I use the same uniform it either updates all the materials or everything gets cached so it uses the first material.

If you use uniform for different materials and values, the final code will be the same, the node hierarchy too, so the system will not need to build and compile a new material/shader.

WestLangley commented on examples/webgpu_tsl_wood.html:
```diff
@@ -0,0 +1,175 @@
+            
+                const directional_light = new THREE.DirectionalLight(0xfdfdfd, 4);
+                directional_light.position.set(-1, 0.3, 0.3).normalize();
+                scene.add(directional_light);
```
I'd suggest removing the directional light.

Optionally, increase `scene.environmentIntensity`, if you feel it is needed. 

Or increase `renderer.toneMappingExposure`.

SeeleyLogan replied:
Lights from the environment shine on top of the wood. I wanted there to be light pointing at the wood in the direction the camera is viewing it from

SeeleyLogan committed:
Tried to fix uniforms

(cherry picked from commit 6fd32bb85bfd8d1e26ab460ceb2f6c987a3e8517)

bhouston committed:
move uniforms.  simplify variable names.  adopt cap-case, format.

(cherry picked from commit ade4e9d6583ee44f3ed80a010b659ac94c92ef42)

bhouston committed:
get rid of randomness

(cherry picked from commit eb5a62f66252956483c1eab700386848dbd4e283)

bhouston commented:
@sunag I've updated the code to try to use uniforms.  I think I did it correctly, but now all the woods are the same. 

<img width="1529" height="945" alt="Screenshot 2025-08-13 at 4 34 12 PM" src="https://github.com/user-attachments/assets/e2559371-84fe-4f7a-9420-7aa3bbca7650" />


(I've also formatted the code and changed variables/method names to be more in line with MrDoob style.)

@WestLangley removed light and increased IBL intensity as well!

SeeleyLogan commented:
@cmhhelgeson I used `wgslFn` for functions I ported from Blender. It was easier to just change some minor syntax than to format each line.

Though I don't see a reason why regular TSL Fn wouldn't work.

bhouston committed:
fix procedural wood using MeshPhysicalNodeMaterial

(cherry picked from commit 9fad1ed547e3812217c4b912327e817718308967)

bhouston committed:
typo

(cherry picked from commit bbd22da0e7a6d6292daff45e50b8558f3fbbab0a)

bhouston committed:
boost HDRI

(cherry picked from commit 7236e5e1054df2a37d5532884d704e3f7caf44ca)

bhouston committed:
allow async loading.

(cherry picked from commit 531d75c2d677ade892f5bf80d2bf75aa9c8d15a1)

SeeleyLogan committed:
Optimizations

(cherry picked from commit dc8562d8f363851c8802940ec7e96ae57ce427ba)

bhouston committed:
fix typos

(cherry picked from commit fdc290ac702799310570e24dd280b7bdea27a00b)

bhouston commented:
@Mugen87



The aliasing is an issue.  We likely have to figure out how to do analytical AA in the shader.  But I figured we could contribute this back and then worry about it?

The flat circles are actually the cellular structure of the wood, or attempting to be.  Here are pictures of real wood from my garage - notice the streaks from the cellular structure - this is maple and walnut, in other woods, they are not quite as prominent:

<img width="583" height="1036" alt="Screenshot 2025-08-14 at 10 31 34 AM" src="https://github.com/user-attachments/assets/75db3c27-1b57-44a8-936d-9745dc2a2d3f" />
<img width="556" height="999" alt="Screenshot 2025-08-14 at 10 31 42 AM" src="https://github.com/user-attachments/assets/c8e8a3a4-eb77-429f-908a-b9a2973ab44c" />

SeeleyLogan committed:
Prettified

(cherry picked from commit e01ab4c3cb37cdee52608dbed37fa8c9b174f45a)

bhouston commented:
Redid the scene and added incremental loading:

![output](https://github.com/user-attachments/assets/fc22b002-1c27-43b1-99d2-cfbefca13012)

SeeleyLogan committed:
Alias Fix

Wood rings and cell structure suffered from aliasing.

Improved by blurring rings based on camera distance and changing cell size based on camera distance.

(cherry picked from commit b49b44bb7f99879b47ce4526580a1fa3711050e4)

SeeleyLogan commented:
@Mugen87 It looks like I've fixed the aliasing issue. As the camera moves away, the sharp wood rings get blurred and the cell structure becomes smaller. There are a couple distances where the anti-aliasing breaks down slightly; nothing too noticeable.

Mugen87 commented:
Wow, much better!

bhouston committed:
tone map exposure to 1

(cherry picked from commit 469157994d0a747763d6cbbccc9d255ce88e7353)

Mugen87 committed:
Update webgpu_tsl_wood.html

Fix code style.

(cherry picked from commit bbbca28e13873d4666b92571d165ca00e66cae6c)

sunag committed:
cache colorNode

(cherry picked from commit 640fca89b9617f23884dd57f61fc6b905ee3ddb9)

sunag committed:
Update webgpu_tsl_wood.html

(cherry picked from commit 37abf5d10f997a87a406361d1a8cbba1026aa7c3)

cmhhelgeson commented:
@sunag The onObjectUpdate just creates a callback that updates the uniform whenever parameters on the material itself are changed, yes?

hybridherbst commented:
I think incremental loading just covers the fact that compilation is slow, I'd opt for removing it and loading everything at the same time (as a typical application would do). 

We're also seeing very slow compile times with TSL (regular MeshPhysicalMaterial is ~50-100x slower to create and compile on WebGPURenderer vs. WebGLRenderer), so I think instead of covering it up this should stay as-is and compilation times hopefully improved at some point in the underlaying system.

bhouston committed:
Update ProceduralWood.js parameters for improved wood texture appearance

(cherry picked from commit 8aa231dd5106b872265a13b4dcc6d999aaec219b)

SeeleyLogan committed:
Converted all instances of wgslFn to Fn

(cherry picked from commit 9119a500a5cc2e9a312cfe9b22c76cd572fafc33)

SeeleyLogan commented:
@cmhhelgeson I've converted all instances of `wgslFn` to proper `Fn`.

SeeleyLogan commented:
@sunag 

> The other point would be to create a WoodMaterial extended material instead of a function, I'm already assigning the parameters to the materials in the function cache I made. So halfway there is already done.

Can you clarify  what you mean by "WoodMaterial extended material instead of a function"?

cmhhelgeson commented:
> Can you clarify what you mean by "WoodMaterial extended material instead of a function"?

ProceduralWood should be a class that extends MeshPhysicalNodeMaterial, rather than a function that creates, modifies, and returns an instance of MeshPhysicalNodeMaterial.

sunag committed:
improve performance - possible cachekey issue

(cherry picked from commit e52cf02c24182781b82f03299f080a742d81e24a)

SeeleyLogan committed:
Created WoodNodeMaterial

Extended StandardMeshNodeMaterial to create WoodNodeMaterial

(cherry picked from commit 26f7487877a812bca95bef88e4e2c2f986bb6d3c)

SeeleyLogan commented:
@sunag Could you help me understand why one `Loop` throws an error and the other doesn't?

```js
TSL.Loop( 3, ( { k } ) => {
		TSL.Loop( 3, ( { j } ) => {
			TSL.Loop( 3, ( { i } ) => {
				let b = TSL.vec3(TSL.float(i).sub(1), TSL.float(j).sub(1), TSL.float(k).sub(1));
                // . . .
```
Throws `THREE.TSL: Invalid parameter for the type "float". [TSLCore.js:728:13](http://localhost:8080/src/nodes/tsl/TSLCore.js)`

```js
TSL.Loop( 3, 3, 3, ( { k, j, i } ) => {
		let b = TSL.vec3(TSL.float(i).sub(1), TSL.float(j).sub(1), TSL.float(k).sub(1));
        // . . .
```

Throws nothing.

Also the error doesn't prevent the shader from compiling.

SeeleyLogan committed:
Reverted voronoi3d back to wgsl

Converting voronoi3d to Fn caused frame drop and input lag. Merging e9f7c8b had no effect.

(cherry picked from commit 72057d2560924f34bed4358664fc22dd2cc94a33)

sunag commented:
> TSL.Loop( 3, ( { k } ) => {

This should work with naming.

```js
Loop( { start: 0, end: 3, name: 'k' }, ( { k } ) => {

} );
```

WestLangley commented:
@SeeleyLogan

https://github.com/mrdoob/three.js/wiki/Three.js-Shading-Language#loop

SeeleyLogan committed:
Added grain direction and made originOffset a uniform

(cherry picked from commit 5ee2693403712033f9d9b1b514ce5d716c46bdbd)

sunag commented on examples/jsm/materials/ProceduralWood.js:
```diff
@@ -0,0 +1,506 @@
+	uniforms.lightGrainColor
+).mul( params.clearcoatDarken );
+
+export class WoodNodeMaterial extends THREE.MeshPhysicalMaterial {
```
MeshPhysicalNodeMaterial

sunag commented:
@SeeleyLogan Something happened in the synchronization, there are many unrelated files changed.

SeeleyLogan committed:
Added back MeshPhysicalNodeMaterial

(cherry picked from commit 895ca13f171d1c3afc5ca26324091158a5101a7d)

mrdoob commented:
This PR is changing 46 files... 🤔
https://github.com/mrdoob/three.js/pull/31640/files

cmhhelgeson commented:
@SeeleyLogan 

I think the issue here is your merge commit [c13dd83], which is pulling in multiple extraneous changes. I would try rebasing your current branch off of dev and force pushing the state of your branch. Personally, I would revert to before the merge commit, rebase my branch, then apply my intended changes, but you can also try rebasing onto dev from the current state of your branch.

```bash
git checkout tsl_wood
git revert <id of commit>
git push 

git rebase dev
// Address any rebase conflicts, which should be minimal or nonexistent in your PR
git push --force-with-lease
// Apply changes made after faulty merge
```

bhouston committed:
originOffset -> grainPosition to align with grainRotation.

bhouston committed:
ensure one can create custom woods, add custom wood block to example.

bhouston committed:
refactor tsl_wood example to clean code

SeeleyLogan committed:
Updated outdated example image

SeeleyLogan committed:
Fixed screenshot size

sunag committed:
add exception

sunag committed:
cleanup

sunag committed:
name `ProceduralWood` -> `WoodNodeMaterial`

Methuselah96 commented on examples/jsm/materials/WoodNodeMaterial.js:
```diff
@@ -0,0 +1,544 @@
+import * as THREE from 'three';
```
Should this be importing from `three/webgpu`?

Methuselah96 commented on examples/jsm/materials/WoodNodeMaterial.js:
```diff
@@ -0,0 +1,544 @@
+		}
+
+		this.colorNode = colorNode;
+		this.clearcoatNode = finalParams.clearcoat;
```
Should this just be called `clearcoat`? It seems like it's not a `Node`.

bhouston commented:
@SeeleyLogan you should check out the project mrdoob did using the wood shader: https://x.com/mrdoob/status/2073666517783458071, live link here: https://mrdoob.github.io/toys/
